### ISSUE TITLE
Low DPI and wrong resolution in Mac OS X High Sierra.

### ISSUE THREAD

**Issue by maeln** [Gargaj/Bonzomatic#53]
I cloned and compiled Bonzomatic on my Mac OS X 10.13.3 and I noticed two bug:
- GLFW seem to use the wrong DPI (the title bar is very blurry).
- The resolution is wrong. A small resolution will end up creating a much bigger window.

Here is a [screenshot](https://maeln.com/img/bonzo-bug-hidpi.png).
I selected a resolution of 1440x900 which is much smaller than my mac resolution (2 880x1 800, 220dpi) and the window is bigger than my screen.

It seem to be related with this [GLFW issue](https://github.com/glfw/glfw/issues/422).
I tried to manually add : 
```
<key>NSHighResolutionCapable</key>
<true/>
```

to the `Info.plist` of the App but the option "Open in low resolution" is still locked to true.

**Comment by maeln** [comment]
Ok, I found the issue:
You need to compile the mac version with the cmake flag `GLFW_USE_RETINA` and then add the thing to the `Info.plist`.

Do you think it's worth mentioning in the README.md in the "Build" section ? I can open a PR.

The font of the editor is still very aliased, but it's more a scintilla issue I suppose.

**Comment by Gargaj** [comment]
@alkama all yours

**Comment by maeln** [comment]
Hello,
Thank you for the quick answer :) .
Adding the flag without messing with the `info.plist` will still make the title bar "blurry", but yes, the actual content of the window remain unchanged.
I will try to use the lastest git version of GLFW to see if I can use it like this for now.

**Comment by alkama** [comment]
Feel free to use that meanwhile:
https://www.dropbox.com/s/m69hrmu3wk3pjhs/Bonzomatic_MacOS_GLFW33.zip?dl=0

**Comment by sacredbanana** [comment]
I don't know if this is related but if I choose 1920x1080 in full screen mode if Bonzomatic is launched from Xcode it doesn't display properly. If I launch Bonzomatic normally it works fine though. Dont know why launching from Xcode would be different.

**Comment by alkama** [comment]
@nkaretnikov the version I compile forcefully disables Retina support because I dont want 4x larger framebuffers than I request, especially not on mac (considering their GPU). Blurry UI elements like title bar is what comes from disabling it.
For the courageous people that want HiDPI reenabled, there's a CMake option that takes care of GLFW part, and the template plist file can be changed for the `NSHighResolutionCapable` field. But last time I tried, on Mojave and Monterey, going fullscreen on HiDPI was completely broken, requested resolution was not provided and framebuffer goes berzerk being either partly offscreen or incorrectly stretched. Since it happens only on newer macos whatever the GLFW version, I tend to think it's a bug there. My debug-printing of "obtained framebuffer sizes" compared to standard requested ones shows totally random numbers.

**Comment by nkaretnikov** [comment]
@alkama fwiw, i don't use fullscreen with the high-res mode enabled because it's messed up indeed. i just tailored the dimensions to match my resolution and do not tick the fullscreen checkbox.

it's not just ui elements like the menu. my main reason for using high-res is font rendering. it looks like it's not aliased at all or maybe expects to be rendered on windows, not sure. e.g., symbols like "v" are not smooth at all. i've tried using bitmap fonts targeting low-res too (like the proggy family), but it's still not good enough

**Comment by alkama** [comment]
@nkaretnikov you might wanna open the console or launch bonzomatic from a terminal to see the lines about requested framebuffer and obtained framebuffer sizes.

They come from https://github.com/Gargaj/Bonzomatic/blob/master/src/platform_glfw/Renderer.cpp#L305-L311

It might reveal some big difference between hiDPI vs non hiDPI support.
Especially considering the fragment shader will run for every pixels of the obtained framebuffer!
With retina support, I observed that when requesting fullhd, instead of having 1920x1080 pixels, I often obtained 3840x2160 instead, which lead to vastly different performance when livecoding :D especially on a mac.

**Comment by nkaretnikov** [comment]
@alkama oh yeah, i noticed that as well.

eg with low-res

```
[GLFW] Requested framebuffer size: 1280 x 840
[GLFW] Obtained framebuffer size: 1280 x 840
```

with high-res (your build + plist change) -- this is in non-fullscreen mode:

```
[GLFW] Requested framebuffer size: 1352 x 792
[GLFW] Obtained framebuffer size: 2704 x 1584
```

**Comment by alkama** [comment]
Yes, that's why I dont promote and actively disabled Retina support on Bonzomatic's default build :D 
It felt safer than having fellow mac users see even less performances than they should just because they render 4x more pixels.

IIRC, shader framebuffer is now offscreen (for feedback), so it could be decoupled from DPI, this would then make it possible to have HiDPI support for text and UI without impacting the shader performance!
Will keep that as a TODO.
